File: checks/ruby/lib/section508/checks/informative_image_alt.rb
# frozen_string_literal: true

module Section508
  module Checks
    # The image has a text alternative that carries the fact.
    class InformativeImageAlt
      WEAK = /^(image|photo|picture|graphic|chart|icon|logo)$/i
      FILENAME = /\.(png|jpe?g|gif|svg|webp)$/i

      def self.call(root)
        return [] unless root

        images = Dom.elements(root, "img")
        return ["This section has no image with a text alternative."] if images.empty?

        images.each do |image|
          return ["This image has no alt attribute."] unless Dom.has_attr?(image, "alt")

          alt = image["alt"].to_s.gsub(/\s+/, " ").strip
          return ["This informative image has empty alt text."] if alt.empty?
          return ["“#{alt}” names the file or format, not the information."] if WEAK.match?(alt) || FILENAME.match?(alt)
        end
        []
      end
    end
  end
end

File: checks/python/src/section508/checks/informative_image_alt.py
import re

from section508.dom import SPACE, attr, has_attr, select

WEAK = re.compile(r"^(image|photo|picture|graphic|chart|icon|logo)$", re.I)
FILENAME = re.compile(r"\.(png|jpe?g|gif|svg|webp)$", re.I)


def check(root):
    if not root:
        return []

    images = select(root, "img")
    if not images:
        return ["This section has no image with a text alternative."]

    for image in images:
        if not has_attr(image, "alt"):
            return ["This image has no alt attribute."]
        alt = SPACE.sub(" ", attr(image, "alt") or "").strip()
        if not alt:
            return ["This informative image has empty alt text."]
        if WEAK.match(alt) or FILENAME.search(alt):
            return [f"“{alt}” names the file or format, not the information."]
    return []
